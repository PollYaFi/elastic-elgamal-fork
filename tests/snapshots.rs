//! Snapshot testing to check compatibility of `serde` serialization of types.

use insta::assert_yaml_snapshot;
use rand_chacha::ChaChaRng;
use rand_core::SeedableRng;

use elastic_elgamal::{
    app::{ChoiceParams, EncryptedChoice},
    group::{Generic, Group, Ristretto},
    Keypair, RangeDecomposition,
};

trait Named {
    const NAME: &'static str;
}

impl Named for Ristretto {
    const NAME: &'static str = "ristretto";
}

impl Named for Generic<::k256::Secp256k1> {
    const NAME: &'static str = "k256";
}

fn test_ciphertext_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let ciphertext = public_key.encrypt(42, &mut rng);

    let full_name = format!("ciphertext-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext);
}

fn test_zero_encryption_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();

    let (ciphertext, proof) = public_key.encrypt_zero(&mut rng);
    let ciphertext_with_proof = serde_json::json!({
        "ciphertext": ciphertext,
        "proof": proof,
    });
    let full_name = format!("zero-encryption-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext_with_proof);
}

fn test_bool_encryption_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();

    let (ciphertext, proof) = public_key.encrypt_bool(true, &mut rng);
    let ciphertext_with_proof = serde_json::json!({
        "ciphertext": ciphertext,
        "proof": proof,
    });
    let full_name = format!("bool-encryption-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext_with_proof);
}

fn test_range_encryption_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let range = RangeDecomposition::optimal(100).into();

    let (ciphertext, proof) = public_key.encrypt_range(&range, 42, &mut rng);
    let ciphertext_with_proof = serde_json::json!({
        "ciphertext": ciphertext,
        "proof": proof,
    });
    let full_name = format!("range-encryption-{}", G::NAME);
    assert_yaml_snapshot!(full_name, ciphertext_with_proof);
}

fn test_encrypted_choice_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let choice_params = ChoiceParams::single(public_key, 5);

    let choice = EncryptedChoice::single(3, &choice_params, &mut rng);
    let full_name = format!("encrypted-choice-{}", G::NAME);
    assert_yaml_snapshot!(full_name, choice);
}

fn test_encrypted_multi_choice_snapshot<G: Group + Named>() {
    let mut rng = ChaChaRng::seed_from_u64(12345);
    let (public_key, _) = Keypair::<G>::generate(&mut rng).into_tuple();
    let choice_params = ChoiceParams::multi(public_key, 5);

    let choices = [false, true, true, false, true];
    let choices = EncryptedChoice::new(&choice_params, &choices, &mut rng);
    let full_name = format!("encrypted-multi-choice-{}", G::NAME);
    assert_yaml_snapshot!(full_name, choices);
}

mod ristretto {
    use super::*;

    #[test]
    fn ciphertext_snapshot() {
        test_ciphertext_snapshot::<Ristretto>();
    }

    #[test]
    fn zero_encryption_snapshot() {
        test_zero_encryption_snapshot::<Ristretto>();
    }

    #[test]
    fn bool_encryption_snapshot() {
        test_bool_encryption_snapshot::<Ristretto>();
    }

    #[test]
    fn range_encryption_snapshot() {
        test_range_encryption_snapshot::<Ristretto>();
    }

    #[test]
    fn encrypted_choice_snapshot() {
        test_encrypted_choice_snapshot::<Ristretto>();
    }

    #[test]
    fn encrypted_multi_choice_snapshot() {
        test_encrypted_multi_choice_snapshot::<Ristretto>();
    }
}

mod k256 {
    use super::*;

    type K256 = Generic<::k256::Secp256k1>;

    #[test]
    fn ciphertext_snapshot() {
        test_ciphertext_snapshot::<K256>();
    }

    #[test]
    fn zero_encryption_snapshot() {
        test_zero_encryption_snapshot::<K256>();
    }

    #[test]
    fn bool_encryption_snapshot() {
        test_bool_encryption_snapshot::<K256>();
    }

    #[test]
    fn range_encryption_snapshot() {
        test_range_encryption_snapshot::<K256>();
    }

    #[test]
    fn encrypted_choice_snapshot() {
        test_encrypted_choice_snapshot::<K256>();
    }

    #[test]
    fn encrypted_multi_choice_snapshot() {
        test_encrypted_multi_choice_snapshot::<K256>();
    }
}
