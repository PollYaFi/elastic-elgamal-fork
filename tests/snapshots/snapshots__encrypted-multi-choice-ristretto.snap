---
source: tests/snapshots.rs
expression: choices

---
variants:
  - random_element: gugt6pqWT9lwtV7WdEdZsVSYTUe0VyNh2w_CzDMKHXQ
    blinded_element: tExVM936V6m9VXTOFajptSiHhVXsac5ZKPMXE6V7vE8
  - random_element: MH-LpdrtV4wiXOhpdvyUnyfwRmCev3ggAcaPzjziZUM
    blinded_element: 5ng8pQx6FK8xJ3Uk7ZrRbsPjxLlejuT_V0Uai0MjGz8
  - random_element: Mmg8vsLhioDg4QxKtuGWXU-FWVTM0OIw9a2DPCi4kw0
    blinded_element: DJFTNS_w8gqyGhpyDhrEtSSmjhI36q9UW6abyJq2n0Q
  - random_element: qOAkbAowCYWYzs-ZQ5r28iXCqk8KOZa2WmwFugs9dGM
    blinded_element: WJKQG1pmYHMkTqA2PmdwgjVZX_S2od0Un4Q63WNqb1M
  - random_element: 3hlVaLYWpp2TqMsOw9QxJpp-KOxYUPX1-X9ZlBdnWms
    blinded_element: rD4ZKNjrlvJy-4SA1xIuRPyjzWkLcdaLZgGKiCmI1y8
range_proof:
  common_challenge: 9W2OaNmOZ4Y9FlqPMcmaKZ46-lvdhHY6OmKyZLk46QM
  ring_responses:
    - QA34-TqArd3cZ4tZ2jJscTb78qkf_3ELvlXx1b8g_g4
    - CoNRRFG5wnN2eagD-EIlsKjTU-bcPvJxao51kp6b2Qk
    - Je2UfVEXzwRNj435C7RLleqRj55GHWtUQENhu5UeswE
    - TfwRWAo3V_tGDR9bUT8qZ9jXHVEABxsn2WfATszReAg
    - OVQmzJrszYnIb0oUFteCjWKpMx_Y2Ew9KFyEAeFejQg
    - uRE_hbZ_G_5PleYKZbo-SmwztlNkbBbNCrXSrN3utg4
    - km21fHo-MMgYXqKjbAvnTuX_jIP-nRMh-syGJD2EHwE
    - HY3H4TRY_Xt5nn1KGaYmoHWfFgmQuMl7TgMwMQj7Agw
    - EIfmWQ_74uv-AbLAn2ifW1sTu2BiaiqYYb4N6az2XAA
    - QH0Q3bMKQ5tek-RB1xf7R7iMe9nJj_Pe_lmMv44igQg
sum_proof: ~

